### DISCUSSION TITLE
What's the best way to send a notification as fast as possible to > 100.000 users?

### DISCUSSION THREAD

**Discussion by cowera** [kreait/firebase-php#629] [General]
I have tried the following:
- sending via topic
Works, but as of late many users complain that they get the notification >7mins after sending it. Including me. Worked for a long time and suddenly notifications got delayed for some reason.
- sending in chunks of 500 to token (Multicast)
Works and I think it gets faster to users than sending via topic, but it takes forever to send and thus it does take minutes - what I would like to achieve is to get this down to seconds!
This is what a message looks like:
```
$not = new stdClass;
        $not->type = "notification";
        $not->link = "https://www.test.com";
        $not->title = "This is a test";
        $not->body = "Sorry, but this is only a test.";

        $topic = "test";

        $notification = NotificationAlias::create(
            $not->title,
            $not->body
        );

$message = CloudMessage::fromArray([
            'priority' => "high",
            'notification' => $notification,
            'data' => (array)json_decode(json_encode($not)),
            'apns' => [
                'headers' => [
                    'apns-priority' => '10',
                ],
                'payload' => [
                    'aps' => [
                        'alert' => [
                            'title' => $not->title,
                            'body' => $not->body
                        ],
                        'sound' => 'default',
                        'mutable-content' => 1,
                        'thread-id' => "topic-" . $topic
                    ],
                ],
            ],
            'android' => [
                'priority' => 'high',
                'notification' => [
                    'default_vibrate_timings' => true,
                    'default_sound' => true,
                    'notification_priority' => 'PRIORITY_HIGH'
                ],
            ],
        ]);
```
I can't emphasize enough how crucial it is for my application that users get the notification immediately. I'd be very grateful for any hints.

**Comment by jeromegamez** [comment]
I think that topic messaging is the most efficient way to do it at the moment - especially since you said that this worked for the longest time and only suddenly got delayed by a lot. It happened before that notifications were delayed and it usually worked itself out after some time. I'd recommend contacting Firebase Support directly to investigate if they know about a current service degradation.

The SDK doesn't support async requests (I thought about: how about sending the chunked batch request in parallel), but I'm pretty sure that Firebase would rate-limit them - the 500 chunk size has certainly been set for a reason 😅

**Reply by cowera** [reply]
Thank you! The problem is that the official advise is kind of confusing:

`Topic messages are optimized for throughput rather than latency. For fast, secure delivery to single devices or small groups of devices, target messages to registration tokens, not topics.`

So on the one hand topics are slow and on the other hand tokens shouldn't be used for larger groups.

edit: I'd really like to try sending the request in parallel, but I'm actually unsure how to approach that using PHP.
